logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/ephemeral/SelfUserReceiverSelfDeletionUseCase.kt

```diff
@@ -0,0 +1,138 @@
+/*
+ * Wire
+ * Copyright (C) 2023 Wire Swiss GmbH
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ */
+package com.wire.kalium.logic.feature.message.ephemeral
+
+import com.benasher44.uuid.uuid4
+import com.wire.kalium.logic.CoreFailure
+import com.wire.kalium.logger.KaliumLogger.Companion.ApplicationFlow.ASSETS
+import com.wire.kalium.logger.KaliumLogger.Companion.ApplicationFlow.MESSAGES
+import com.wire.kalium.logic.cache.SelfConversationIdProvider
+import com.wire.kalium.logic.data.asset.AssetRepository
+import com.wire.kalium.logic.data.conversation.ClientId
+import com.wire.kalium.logic.data.id.ConversationId
+import com.wire.kalium.logic.data.message.Message
+import com.wire.kalium.logic.data.message.MessageContent
+import com.wire.kalium.logic.data.message.MessageRepository
+import com.wire.kalium.logic.data.user.UserId
+import com.wire.kalium.logic.feature.CurrentClientIdProvider
+import com.wire.kalium.logic.feature.message.MessageSender
+import com.wire.kalium.logic.functional.Either
+import com.wire.kalium.logic.functional.flatMap
+import com.wire.kalium.logic.functional.foldToEitherWhileRight
+import com.wire.kalium.logic.functional.map
+import com.wire.kalium.logic.functional.onFailure
+import com.wire.kalium.logic.functional.onSuccess
+import com.wire.kalium.logic.kaliumLogger
+import com.wire.kalium.util.DateTimeUtil
+
+/**
+ * When the self user is receiver of the self deletion message,
+ * we delete it permanently after expiration and inform the sender by broadcasting a message to delete
+ * for the self-deleting message, before the receiver does it on the sender side, the message is simply marked as deleted
+ * see [com.wire.kalium.logic.feature.message.ephemeral.SelfUserReceiverSelfDeletionUseCase]
+ **/
+internal class SelfUserReceiverSelfDeletionUseCase(
+    private val messageRepository: MessageRepository,
+    private val assetRepository: AssetRepository,
+    private val currentClientIdProvider: CurrentClientIdProvider,
+    private val messageSender: MessageSender,
+    private val selfUserId: UserId,
+    private val selfConversationIdProvider: SelfConversationIdProvider
+) {
```

Comment:
Tests here bitte.

Reply:
Test will be coming after this solution is approved